File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/common/template/FileBasedPromptTemplate.java
package com.study.webflux.rag.infrastructure.common.template;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;

import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Component;
import org.springframework.util.StreamUtils;

@Component
public class FileBasedPromptTemplate {

	private static final List<String> TEMPLATE_EXTENSIONS = List.of(".md", ".txt");

	public String load(String templateName) {
		return load(templateName, Map.of());
	}

	public String load(String templateName, Map<String, String> variables) {
		try {
			ClassPathResource resource = resolveResource(templateName);
			String template = StreamUtils.copyToString(resource.getInputStream(),
				StandardCharsets.UTF_8);

			String result = template;
			for (Map.Entry<String, String> entry : variables.entrySet()) {
				result = result.replace("{{" + entry.getKey() + "}}", entry.getValue());
			}
			return result;
		} catch (IOException e) {
			throw new RuntimeException("Failed to load template: " + templateName, e);
		}
	}

	private ClassPathResource resolveResource(String templateName) throws IOException {
		for (String ext : TEMPLATE_EXTENSIONS) {
			ClassPathResource resource = new ClassPathResource("templates/" + templateName + ext);
			if (resource.exists()) {
				return resource;
			}
		}
		throw new IOException("Template not found for name: " + templateName);
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/infrastructure/common/template/FileBasedPromptTemplate.kt
package com.miyou.app.infrastructure.common.template

import com.miyou.app.domain.dialogue.port.TemplateLoaderPort
import org.springframework.core.io.ClassPathResource
import org.springframework.stereotype.Component
import org.springframework.util.StreamUtils
import java.io.IOException
import java.nio.charset.StandardCharsets

@Component
class FileBasedPromptTemplate : TemplateLoaderPort {
    private val templateExtensions: List<String> = listOf(".md", ".txt")

    override fun load(templateName: String): String = load(templateName, emptyMap())

    fun load(
        templateName: String,
        variables: Map<String, String>,
    ): String {
        try {
            val resource = resolveResource(templateName)
            val template = StreamUtils.copyToString(resource.inputStream, StandardCharsets.UTF_8)

            var result = template
            for ((key, value) in variables) {
                result = result.replace("{{$key}}", value)
            }
            return result
        } catch (e: IOException) {
            throw RuntimeException("Failed to load template: $templateName", e)
        }
    }

    private fun resolveResource(templateName: String): ClassPathResource {
        for (ext in templateExtensions) {
            val resource = ClassPathResource("templates/$templateName$ext")
            if (resource.exists()) {
                return resource
            }
        }
        throw IOException("Template not found for name: $templateName")
    }
}
